src/rust-crypto/backup.ts

```diff
@@ -361,29 +360,38 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
-                // suddenly become very slow (40s, 60s or more). Even on other configurations, the count can take several seconds.
-                // This will block other operations on the database, like sending messages
-                // This is a workaround to avoid calling `olmMachine.roomKeyCounts()` too often, and only when necessary.
-                // We don't call it on the first loop because there could be only a few keys to upload, and we don't want to wait for the count.
-                if (!isFirstIteration && remainingToUploadCount === null) {
-                    try {
-                        const keyCount = await this.olmMachine.roomKeyCounts();
-                        remainingToUploadCount = keyCount.total - keyCount.backedUp;
-                    } catch (err) {
-                        logger.error("Backup: Failed to get key counts from rust crypto-sdk", err);
-                    }
-                }
-
                 try {
                     await this.outgoingRequestProcessor.makeOutgoingRequest(request);
                     numFailures = 0;
                     if (this.stopped) break;
+
+                    // Key count performance (`olmMachine.roomKeyCounts()`) can be pretty bad on some configurations.
+                    // In particular, we detected on some M1 macs that when the object store reaches a threshold, the count
+                    // performance stops growing in O(n) and suddenly becomes very slow (40s, 60s or more).
+                    // For reference, the performance drop occurs around 300-400k keys on the platforms where this issue is observed.
+                    // Even on other configurations, the count can take several seconds.
+                    // This will block other operations on the database, like sending messages.
+                    //
+                    // This is a workaround to avoid calling `olmMachine.roomKeyCounts()` too often, and only when necessary.
+                    // We don't call it on the first loop because there could be only a few keys to upload, and we don't want to wait for the count.
+                    if (!isFirstIteration && remainingToUploadCount === null) {
+                        try {
+                            const keyCount = await this.olmMachine.roomKeyCounts();
+                            remainingToUploadCount = keyCount.total - keyCount.backedUp;
+                        } catch (err) {
+                            logger.error("Backup: Failed to get key counts from rust crypto-sdk", err);
+                        }
+                    }
+
                     if (remainingToUploadCount !== null) {
+                        this.emit(CryptoEvent.KeyBackupSessionsRemaining, remainingToUploadCount);
                         const keysCountInBatch = this.keysCountInBatch(request);
+                        // The `remainingToUploadCount` is computed only once for the current backupKeysLoop. But new
```

Comment:
This isn't accurate; we compute `remainingToUploadCount` on each iteration.

```suggestion
                        // `OlmMachine.roomKeyCounts` is called only once for the current backupKeysLoop. But new
```